packages/zql/src/ivm/flipped-join.ts

```diff
@@ -0,0 +1,440 @@
+ * stream of related nodes from the child input (in child input order).
+ */
+export class FlippedJoin implements Input {
+  readonly #parent: Input;
+  readonly #child: Input;
+  readonly #parentKey: CompoundKey;
+  readonly #childKey: CompoundKey;
+  readonly #relationshipName: string;
+  readonly #schema: SourceSchema;
+
+  #output: Output = throwOutput;
+
+  #inprogressChildChange: JoinChangeOverlay | undefined;
+
+  constructor({
+    parent,
+    child,
+    parentKey,
+    childKey,
+    relationshipName,
+    hidden,
+    system,
+  }: Args) {
+    assert(parent !== child, 'Parent and child must be different operators');
+    assert(
+      parentKey.length === childKey.length,
+      'The parentKey and childKey keys must have same length',
+    );
+    this.#parent = parent;
+    this.#child = child;
+    this.#parentKey = parentKey;
+    this.#childKey = childKey;
+    this.#relationshipName = relationshipName;
+
+    const parentSchema = parent.getSchema();
+    const childSchema = child.getSchema();
+    this.#schema = {
+      ...parentSchema,
+      relationships: {
+        ...parentSchema.relationships,
+        [relationshipName]: {
+          ...childSchema,
+          isHidden: hidden,
+          system,
+        },
+      },
+    };
+
+    parent.setOutput({
+      push: (change: Change) => this.#pushParent(change),
+    });
+    child.setOutput({
+      push: (change: Change) => this.#pushChild(change),
+    });
+  }
+
+  destroy(): void {
+    this.#child.destroy();
+    this.#parent.destroy();
+  }
+
+  setOutput(output: Output): void {
+    this.#output = output;
+  }
+
+  getSchema(): SourceSchema {
+    return this.#schema;
+  }
+
+  // TODO: When parentKey is the parent's primary key (or more
+  // generally when the parent cardinality is expected to be small) a different
+  // algorithm should be used:  For each child node, fetch all parent nodes
+  // eagerly and then sort using quicksort.
+  *fetch(req: FetchRequest): Stream<Node> {
+    const childNodes = [...this.#child.fetch({})];
+    // FlippedJoin's split-push change overlay logic is largely
+    // the same as Join's, which the exception of remove.  For remove,
+    // the change is undone here, and then re-applied to parents with order
+    // less than or equal to change.position below.  This is necessary
+    // because if the removed node was the last related child, the
+    // related parents with position greater than change.position
+    // (which should not yet have the node removed), would not even
+    // be fetched here, and would be absent from the output all together.
+    if (this.#inprogressChildChange?.change.type === 'remove') {
+      const removedNode = this.#inprogressChildChange.change.node;
+      const compare = this.#child.getSchema().compareRows;
+      const insertPos = binarySearch(childNodes.length, i =>
+        compare(childNodes[i].row, removedNode.row),
+      );
```

Comment:
`binarySearch` looks like it expects `needle - haystack[i]` not `haystack[i] - needle`

https://github.com/rocicorp/mono/blob/345895c186e1f0eaa8b4a5f2d4b7e5e678f592e3/packages/shared/src/binary-search.test.ts#L8


```suggestion
      const insertPos = binarySearch(childNodes.length, i =>
        compare(removedNode.row, childNodes[i].row),
      );
```